### PR TITLE
Fix SynthesizedFileUnit serialization and TBDGen issues.

### PR DESCRIPTION
**Pull request by dan-zheng**
Fix `SynthesizedFileUnit` serialization and TBDGen issues exposed in TF-1239.

Make `SynthesizedFileUnit` attached to a `SourceFile`. This seemed like the
least ad-hoc approach to avoid doing unnecessary work for other `FileUnit`s.

TBDGen: when visiting a `SourceFile`, also visit its `SynthesizedFileUnit` if
it exists.

Serialization: do not treat `SynthesizedFileUnit` declarations as xrefs when
serializing the companion `SourceFile`.

---

Upstream AutoDiff tests validating that TF-1239 is fixed.
Unblocks end-to-end differentiation, particularly cross-module compilation use cases.

### CODE REVIEW

**Comment by dan-zheng** [comment]
This continues `SynthesizedFileUnit` added in https://github.com/apple/swift/pull/30845 based on [feedback here](https://github.com/apple/swift/pull/30845#issuecomment-610145488).

Some context: the SILOptimizer differentiation transform generates both struct and enum declarations and adds them to a `SynthesizedFileUnit`. To avoid generating declarations in SILOptimizer: struct declarations could be replaced with tuples, at the cost of gravely hurting SIL readability. But enum declarations are not easily replaceable.

Related discussion about generating declarations in SILOptimizer: https://github.com/apple/swift/pull/30736#issuecomment-607720082

---

Though the old approach of "mutating an existing `SourceFile`" was simpler, it's not clear to me that `SynthesizedFileUnit` is worth the complexity.

**Comment by dan-zheng** [comment]
@swift-ci Please test

File: include/swift/AST/SourceFile.h [hunk 1]
```diff
@@ -14,11 +14,13 @@
 #define SWIFT_AST_SOURCEFILE_H
 
 #include "swift/AST/FileUnit.h"
+#include "swift/AST/SynthesizedFileUnit.h"
 #include "swift/Basic/Debug.h"
 
 namespace swift {
 
 class PersistentParserState;
+class SynthesizedFileUnit;
```
Comment (dan-zheng) [inline include/swift/AST/SourceFile.h:23]: This line can be removed. I'll remove it in a follow-up PR, unless I need to update this one.

**Review by rxwei — APPROVED** [review]

**Comment by dan-zheng** [comment]
Merging to unblock progress. Happy to address any feedback later!

File: include/swift/AST/SourceFile.h (hunk 1 above)
Comment (dan-zheng) [reply include/swift/AST/SourceFile.h:23]: Done in https://github.com/apple/swift/pull/30917.

**Pull request merged**